src/nucleus/tt.ml

```diff
@@ -135,10 +133,12 @@ and instantiate es ?(lvl=0) ({term=e';assumptions;loc;} as e) =
   in
   match e' with
 
-    | Type -> {e with assumptions}
+    | Type -> {e with assumptions} (* XXX why do we need assumptions here? Type is well formed by itself *)
```

Comment:
Assumptions are on terms because they're used to get the subterm judgement. So what actually happens is we have `f : B -> foo` and `eq : A == B |- f a : foo` (resp `eq : Type == B |- f Type : foo`), and the `a` subterm (resp `Type` subterm) is where `eq` needs to be mentioned.

If we only produce terms and never destruct them (with matching) then the context is minimal. However when we match we do need to restrict to get a minimal context. (eg `match assume a : A in assume b : B in pair A B a b with pair _ _ ?x _ => ... end` currently we bind `x <- a : A, b : B |- a : A`).
Substitution might produce too-big contexts as well, I'm not sure.
There's a choice to be made between manual restrict calls from user mode versus automatically restricting when there's a chance a non minimal judgement would be produced. Currently we have the first option with the call to restrict only implemented in my unmerged branch.

I should probably write up how assumptions work in the doc to make it clearer and so random people can look at the repository and not need to ask what's going on.